### PR TITLE
feat: expose API for upgrading from basic to x509 credentials for a conversation WPB-10161

### PR DESCRIPTION
**Pull request by typfel**
# What's new in this PR

Expose API for upgrading from basic to x509 credentials for a conversation. This is useful for a clients to address the race condition, which can happen if they got added to a conversation with their basic credentials while upgrading to x509 credentials. 

----
##### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: crypto-ffi/bindings/js/CoreCrypto.ts [hunk 1]
```diff
@@ -1559,6 +1559,47 @@ export class CoreCrypto {
...
+                welcome: ffiRet.welcome,
+                commit: ffiRet.commit,
+                groupInfo: {
+                    encryptionType: gi.encryption_type,
+                    ratchetTreeType: gi.ratchet_tree_type,
+                    payload: gi.payload,
+                },
+            };
+
+            return ret;
```
Comment (SimonThormeyer) [inline crypto-ffi/bindings/js/CoreCrypto.ts:1597]: Let's return this directly without the intermediate variable :)

**Review by SimonThormeyer — APPROVED** [review]
Just a nitpick, otherwise it looks good!

File: crypto-ffi/bindings/js/CoreCrypto.ts (hunk 1 above)
Comment (typfel) [reply crypto-ffi/bindings/js/CoreCrypto.ts:1597]: Sure I could but all other functions use this pattern. I'd say either I change all functions or do nothing.

Comment (typfel) [reply crypto-ffi/bindings/js/CoreCrypto.ts]: Alright went a head updated all functions

File: crypto/src/e2e_identity/rotate.rs [hunk 2]
```diff
@@ -1067,5 +1067,89 @@ pub(crate) mod tests {
...
+                            let commit = alice_central.mls_central.e2ei_rotate(&id, None).await.unwrap();
+
+                            // Bob decrypts the commit...
+                            let decrypted = bob_central
+                                .mls_central
+                                .decrypt_message(&id, commit.commit.to_bytes().unwrap())
+                                .await
+                                .unwrap();
+                            // ...and verifies that now Alice is represented with her new identity
+                            alice_central.mls_central.verify_sender_identity(&case, &decrypted);
```
Comment (istankovic) [inline crypto/src/e2e_identity/rotate.rs:1132]: Hm, here it is Alice, not Bob that is doing sender verification -- is this intentional?

File: crypto/src/e2e_identity/rotate.rs [hunk 3]
```diff
@@ -1067,5 +1067,89 @@ pub(crate) mod tests {
...
+                                .verify_local_credential_rotated(&id, new_handle, new_display_name)
+                                .await;
+
+                            let final_count = alice_central.mls_central.count_entities().await;
+                            assert_eq!(init_count.encryption_keypair, final_count.encryption_keypair);
+                            assert_eq!(
+                                init_count.epoch_encryption_keypair,
+                                final_count.epoch_encryption_keypair
+                            );
+                            assert_eq!(init_count.key_package, final_count.key_package);
```
Comment (istankovic) [inline crypto/src/e2e_identity/rotate.rs:1147]: This means a search for a basic credential key package would fail now, right?

File: crypto-ffi/bindings/gradle/libs.versions.toml [hunk 4]
```diff
@@ -11,7 +11,7 @@ android-junit = "1.1.5"
 android-logback = "2.0.0"
 android-tools = "8.0.0"
 sdk-compile = "33"
-sdk-min = "26"
+sdk-min = "33"
```
Comment (istankovic) [inline crypto-ffi/bindings/gradle/libs.versions.toml:14]: Why this change? If this is unrelated to the other changes in this commit, I'd prefer if this was moved into its own, separate commit, together with a message describing the "why".

**Comment by istankovic** [comment]
The hunk(s) modifying crypto-ffi/src/generic.rs in commit 961e61a368dca1c1c76392377ebf5fdf695ad5bf should really be incorporated into commit b8b5ef92c1dbde014afa817e84888050ca849896, which introduces that code.

File: crypto-ffi/bindings/gradle/libs.versions.toml (hunk 4 above)
Comment (typfel) [reply crypto-ffi/bindings/gradle/libs.versions.toml:14]: When building the generated android bindings the compiler said it was using things SDK level 33. But you're I'll alright try to reproduce this on main and make separate PR for it.

File: crypto-ffi/bindings/jvm/src/main/kotlin/com/wire/crypto/client/MLSClient.kt [hunk 5]
```diff
@@ -307,6 +307,20 @@ class MLSClient(private val cc: com.wire.crypto.CoreCrypto) {
...
     suspend fun updateKeyingMaterial(id: MLSGroupId) = cc.updateKeyingMaterial(id.lower()).lift()
 
+    /**
+     * Creates an update commit which replaces your leaf containing basic credentials with a leaf node containing x509 credentials in the conversation.
+     *
+     * NOTE: you can only call this after you've completed the enrollment for an end-to-end identity, calling this without
+     * a valid end-to-end identity will result in an error.
+     *
+     * **CAUTION**: [commitAccepted] **HAS TO** be called afterward **ONLY IF** the Delivery Service responds'200 OK' to the [CommitBundle] upload.
+     * It will "merge" the commit locally i.e. increment the local group epoch, use new encryption secrets etc...
```
[the remark this answers is not in the document]
Comment (typfel) [reply crypto-ffi/bindings/jvm/src/main/kotlin/com/wire/crypto/client/MLSClient.kt:317]: Sure, I was just copy pasted from other similar methods which also generate commit bundles though.

File: crypto/src/e2e_identity/rotate.rs (hunk 2 above)
Comment (typfel) [reply crypto/src/e2e_identity/rotate.rs:1132]: Hmm, I inherited it from the test I based this test on but it looks wrong. I'll investigate

File: crypto/src/e2e_identity/rotate.rs (hunk 3 above)
Comment (typfel) [reply crypto/src/e2e_identity/rotate.rs:1147]: I mean this count doesn't distinguish. In this test the key packages aren't touched. 10 key packages are generated when the centrals are created but that's it.

I honestly now that I'm looking at it I don't see the point of asserting that the `key_package` or `epoch_encryption_keypair` count hasn't changed.

File: crypto/src/e2e_identity/rotate.rs (hunk 2 above)
Comment (typfel) [reply crypto/src/e2e_identity/rotate.rs:1132]: So yes, because `verify_sender_identity` verifies that the sender identity present in the decrypted message matches the identity in the central's credential bundle. 

Since it was Alice who sent the message, `verify_sender_identity` must be called on Alice's central.

File: crypto/src/e2e_identity/rotate.rs (hunk 3 above)
Comment (istankovic) [reply crypto/src/e2e_identity/rotate.rs:1147]: Yeah, then I would omit those checks in order to avoid confusion. :+1:

File: crypto-ffi/bindings/gradle/libs.versions.toml (hunk 4 above)
Comment (typfel) [reply crypto-ffi/bindings/gradle/libs.versions.toml:14]: Ok, we should definitely not raise the `sdk-min` to 33.

This comes from https://github.com/mozilla/uniffi-rs/issues/2226

I'll fix this in separate PR, the cleaner implementation is supposed to fallback to something else on older phones. I'll investigate a bit more if this is just linter issue or coming from the fact that we share code between android and java.

Comment (typfel) [reply crypto-ffi/bindings/gradle/libs.versions.toml]: Ok, reverted this and issue was fixed in https://github.com/wireapp/core-crypto/pull/715

**Review by istankovic — APPROVED** [review]
